src/rpc/blockchain.cpp

```diff
@@ -1612,6 +1613,303 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+          : "No such mempool transaction. Use -txindex to enable blockchain transaction queries"));
+    }
+    return tx_out;
+}
+
+template<typename T>
+static T CalculateTruncatedMedian(std::vector<T>& scores)
+{
+    size_t size = scores.size();
+    if (size == 0) {
+        return 0;
+    }
+
+    std::sort(scores.begin(), scores.end());
+    if (size % 2 == 0) {
+        return (scores[size / 2 - 1] + scores[size / 2]) / 2;
+    } else {
+        return scores[size / 2];
+    }
+}
+
+// outpoint (needed for the utxo index) + nHeight + fCoinBase
+static const size_t PER_UTXO_OVERHEAD = sizeof(COutPoint) + sizeof(uint32_t) + sizeof(bool);
+
+static std::map<std::string, UniValue> GetBlockStatsMap(const CBlock& block, const CBlockIndex* pindex, const std::set<std::string>& stats)
+{
+    std::map<std::string, UniValue> map_stats;
+    int64_t inputs = 0;
+    int64_t outputs = 0;
+    int64_t swtxs = 0;
+    int64_t total_size = 0;
+    int64_t total_weight = 0;
+    int64_t swtotal_size = 0;
+    int64_t swtotal_weight = 0;
+    int64_t utxo_size_inc = 0;
+    CAmount total_out = 0;
+    CAmount totalfee = 0;
+    CAmount minfee = MAX_MONEY;
+    CAmount maxfee = 0;
+    CAmount minfeerate = MAX_MONEY;
+    CAmount maxfeerate = 0;
+    std::vector<CAmount> fee_array;
+    std::vector<CAmount> feerate_array;
+    int64_t mintxsize = MAX_BLOCK_SERIALIZED_SIZE;
+    int64_t maxtxsize = 0;
+    std::vector<int64_t> txsize_array;
+
+    const bool do_mediantxsize = stats.count("mediantxsize") != 0;
+    const bool do_medianfee = stats.count("medianfee") != 0;
+    const bool do_medianfeerate = stats.count("medianfeerate") != 0;
+
+    const bool loop_outputs = stats.count("total_out") != 0 || stats.count("utxo_size_inc") != 0 ||
+        stats.count("totalfee") != 0 || stats.count("avgfee") != 0 || stats.count("avgfeerate") != 0 ||
+        do_medianfee || stats.count("minfee") != 0 || stats.count("maxfee") != 0 ||
+        do_medianfeerate || stats.count("minfeerate") != 0 || stats.count("maxfeerate") != 0;
+
+    const bool do_calculate_size = do_mediantxsize || stats.count("total_size") != 0 || stats.count("avgtxsize") != 0 ||
+        stats.count("mintxsize") != 0 || stats.count("maxtxsize") != 0 || stats.count("swtotal_size") != 0;
+
+    const bool do_calculate_weight = stats.count("total_weight") != 0 || stats.count("avgfeerate") != 0 ||
+        stats.count("swtotal_weight") != 0 || stats.count("medianfeerate_weight") != 0;
+
+    const bool do_calculate_sw = stats.count("swtxs") != 0 || stats.count("swtotal_size") != 0 ||
+        stats.count("swtotal_weight") != 0;
+
+    const bool loop_inputs = stats.count("totalfee") != 0 || stats.count("avgfee") != 0 || stats.count("avgfeerate") != 0 ||
+        do_medianfee || stats.count("minfee") != 0 || stats.count("maxfee") != 0 ||
+        do_medianfeerate || stats.count("minfeerate") != 0 || stats.count("maxfeerate") != 0;
```

Comment:
`loop_inputs` is also needed when `utxo_size_inc` is requested. Setting `loop_outputs = loop_inputs || stats.count("total_out")` might simplify things a little.